ipn/ipnext/ipnext.go

```diff
@@ -0,0 +1,284 @@
+	Profiles() ProfileServices
+
+	// RegisterAuditLogProvider registers an audit log provider,
+	// which returns a function to be called when an auditable action
+	// is about to be performed. The returned function unregisters the provider.
+	// It is a runtime error to register a nil provider.
+	RegisterAuditLogProvider(AuditLogProvider) (unregister func())
+
+	// AuditLogger returns a function that calls all currently registered audit loggers.
+	// The function fails if any logger returns an error, indicating that the action
+	// cannot be logged and must not be performed.
+	//
+	// The returned function captures the current state (e.g., the current profile) at
+	// the time of the call and must not be persisted.
+	AuditLogger() ipnauth.AuditLogFunc
+
+	// RegisterControlClientCallback registers a function to be called every time a new
+	// control client is created. The returned function unregisters the callback.
+	// It is a runtime error to register a nil callback.
+	RegisterControlClientCallback(NewControlClientCallback) (unregister func())
+}
+
+// ProfileServices provides access to the [Host]'s profile management services,
+// such as switching profiles and registering profile change callbacks.
+type ProfileServices interface {
+	// SwitchToBestProfileAsync asynchronously selects the best profile to use
+	// and switches to it, unless it is already the current profile.
+	//
+	// If an extension needs to know when a profile switch occurs,
+	// it must use [ProfileServices.RegisterProfileChangeCallback]
+	// to register a [ProfileChangeCallback].
+	//
+	// The reason indicates why the profile is being switched, such as due
+	// to a client connecting or disconnecting or a change in the desktop
+	// session state. It is used for logging.
+	SwitchToBestProfileAsync(reason string)
+
+	// RegisterBackgroundProfileResolver registers a function to be used when
+	// resolving the background profile. The returned function unregisters the resolver.
+	// It is a runtime error to register a nil resolver.
+	//
+	// TODO(nickkhyl): allow specifying some kind of priority/altitude for the resolver.
+	// TODO(nickkhyl): make it a "profile resolver" instead of a "background profile resolver".
+	// The concepts of the "current user", "foreground profile" and "background profile"
+	// only exist on Windows, and we're moving away from them anyway.
+	RegisterBackgroundProfileResolver(ProfileResolver) (unregister func())
+
+	// RegisterProfileChangeCallback registers a function to be called when the current
+	// [ipn.LoginProfile] changes. The returned function unregisters the callback.
+	// It is a runtime error to register a nil callback.
+	RegisterProfileChangeCallback(ProfileChangeCallback) (unregister func())
+}
+
+// ProfileStore provides read-only access to available login profiles and their preferences.
+// It is not safe for concurrent use and can only be used from the callback it is passed to.
+type ProfileStore interface {
```

Comment:
Something I just thought of: Does any of this refactoring (or the stuff from yesterday) affect our profile store querying in `tsca.dll`?